intake_esm/core.py

```diff
@@ -58,31 +57,201 @@ class esm_datastore(intake.catalog.Catalog):
+        self.aggregation_info = self._get_aggregation_info()
         self._entries = {}
-        self._ds = None
-        self.zarr_kwargs = None
-        self.cdf_kwargs = None
-        self.preprocess = None
-        self.aggregate = None
+        self._grouped = self.df.groupby(self.aggregation_info['groupby_attrs'])
+        self._keys = list(self._grouped.groups.keys())
+        super(esm_datastore, self).__init__(**kwargs)
+
+    def _get_aggregation_info(self):
+        groupby_attrs = []
+        data_format = None
+        format_column_name = None
+        variable_column_name = None
+        aggregations = []
+        aggregation_dict = {}
+        agg_columns = []
+        path_column_name = self.esmcol_data['assets']['column_name']
+
+        if 'format' in self.esmcol_data['assets']:
+            data_format = self.esmcol_data['assets']['format']
+        else:
+            format_column_name = self.esmcol_data['assets']['format_column_name']
+
+        if 'aggregation_control' in self.esmcol_data:
+            aggregation_dict = {}
+            variable_column_name = self.esmcol_data['aggregation_control']['variable_column_name']
+            groupby_attrs = self.esmcol_data['aggregation_control'].get('groupby_attrs', [])
+            aggregations = self.esmcol_data['aggregation_control'].get('aggregations', [])
+            # Sort aggregations to make sure join_existing is always done before join_new
+            aggregations = sorted(aggregations, key=lambda i: i['type'], reverse=True)
+            for agg in aggregations:
+                key = agg['attribute_name']
+                rest = agg.copy()
+                del rest['attribute_name']
+                aggregation_dict[key] = rest
+            agg_columns = list(aggregation_dict.keys())
+
+        if not groupby_attrs:
+            groupby_attrs = self.df.columns.tolist()
+
+        # filter groupby_attrs to ensure no columns with all nans
+        def _allnan_or_nonan(column):
+            if self.df[column].isnull().all():
+                return False
+            elif self.df[column].isnull().any():
+                raise ValueError(
+                    f'The data in the {column} column should either be all NaN or there should be no NaNs'
+                )
+            else:
+                return True
+
+        groupby_attrs = list(filter(_allnan_or_nonan, groupby_attrs))
+
+        info = {
+            'groupby_attrs': groupby_attrs,
+            'variable_column_name': variable_column_name,
+            'aggregations': aggregations,
+            'agg_columns': agg_columns,
+            'aggregation_dict': aggregation_dict,
+            'path_column_name': path_column_name,
+            'data_format': data_format,
+            'format_column_name': format_column_name,
+        }
+        return info
+
+    @property
+    def keys(self):
+        keys = list(map(lambda x: self.sep.join(x), self._keys))
+        return keys
+
+    def _load(self):
```

Comment:
Constructing intake catalog entries appears to be very expensive:

<img width="1499" alt="Screen Shot 2020-04-07 at 2 55 18 PM" src="https://user-images.githubusercontent.com/13301940/78718765-16dfe200-78e0-11ea-95e8-a7fb63442085.png">

This is a huge bottleneck for large intake-esm catalog. 

@jhamman, let me know if you have any ideas on how to proceed.

Reply:
@andersy005 - can you show a similar profile for a single call to `LocalCatalogEntry`?

Reply:
>  can you show a similar profile for a single call to LocalCatalogEntry?

Here's one:

<img width="1259" alt="Screen Shot 2020-04-07 at 6 36 46 PM" src="https://user-images.githubusercontent.com/13301940/78732064-cb88fc00-78fe-11ea-9fbc-15d2e7485a84.png">
<img width="1343" alt="Screen Shot 2020-04-07 at 6 36 58 PM" src="https://user-images.githubusercontent.com/13301940/78732071-d04db000-78fe-11ea-9353-73517bf08f24.png">

Reply:
@andersy005 - thanks but I think we're missing some of the screen shot.

Reply:
I omitted the doc string part:

<img width="1418" alt="Screen Shot 2020-04-07 at 10 53 41 PM" src="https://user-images.githubusercontent.com/13301940/78745872-a60de980-7922-11ea-8573-2760eae2f3ac.png">

Reply:
Thinking about this a bit, I guess I'm not particularly surprised. We're creating 7k+ `LocalCatalogEntries`. Is the data source pulling some metadata at this stage? Would it be reasonable to make the construction of these entries lazy?

Reply:
Thank you for the suggestion! Making the construction lazy improves the performance:

```python
In [1]: import intake                                                       

In [2]: url = "https://raw.githubusercontent.com/NCAR/intake-esm-datastore/m
   ...: aster/catalogs/pangeo-cmip6.json"                                   

In [3]: %%time 
   ...: col = intake.open_esm_datastore(url) 
   ...:  
   ...:                                                                     
CPU times: user 1.87 s, sys: 283 ms, total: 2.15 s
Wall time: 5.88 s
```

This was taking more than 60 seconds before making this change. 


<img width="906" alt="Screen Shot 2020-04-08 at 11 38 12 AM" src="https://user-images.githubusercontent.com/13301940/78815444-87910800-798d-11ea-9010-bc71eb35c434.png">